File: go/internal/tools/echo.go
package tools

import (
	"encoding/json"
	"fmt"
	"time"
)

// EchoInput represents input for echo tool
type EchoInput struct {
	Message string `json:"message"`
}

// Echo returns the input message
func Echo(args map[string]interface{}) (interface{}, error) {
	argBytes, err := json.Marshal(args)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal arguments: %w", err)
	}

	var input EchoInput
	if err := json.Unmarshal(argBytes, &input); err != nil {
		return nil, fmt.Errorf("failed to unmarshal arguments: %w", err)
	}

	if input.Message == "" {
		return nil, fmt.Errorf("message is required")
	}

	return map[string]interface{}{
		"echo":      input.Message,
		"length":    len(input.Message),
		"timestamp": time.Now().Format(time.RFC3339),
	}, nil
}

// GetEchoSchema returns the JSON schema for echo
func GetEchoSchema() map[string]interface{} {
	return map[string]interface{}{
		"type": "object",
		"properties": map[string]interface{}{
			"message": map[string]interface{}{
				"type":        "string",
				"description": "The message to echo back",
			},
		},
		"required": []string{"message"},
	}
}

File: rust/mcp-core/src/tools/echo.rs
use super::ToolHandler;
use crate::types::Tool;
use anyhow::{anyhow, Result};
use serde::Deserialize;
use serde_json::{json, Value};
use std::collections::HashMap;

#[derive(Debug, Deserialize)]
struct EchoInput {
    message: String,
}

pub struct EchoTool;

#[async_trait::async_trait]
impl ToolHandler for EchoTool {
    fn definition(&self) -> Tool {
        Tool {
            name: "echo".to_string(),
            description: "Echo back a message (for testing)".to_string(),
            input_schema: json!({
                "type": "object",
                "properties": {
                    "message": {
                        "type": "string",
                        "description": "The message to echo back"
                    }
                },
                "required": ["message"]
            }),
        }
    }

    async fn execute(&self, args: HashMap<String, Value>) -> Result<Value> {
        let input: EchoInput = serde_json::from_value(json!(args))?;

        if input.message.is_empty() {
            return Err(anyhow!("Message is required"));
        }

        Ok(json!({
            "echo": input.message,
            "length": input.message.len(),
            "timestamp": chrono::Utc::now().to_rfc3339(),
        }))
    }
}
